src/lib/Embed.hs

```diff
@@ -736,6 +781,37 @@ intToIndexE ty@(TC con) i = case con of
     return $ Con $ SumCon isLeft li ri
   _ -> error $ "Unexpected type " ++ pprint con
   where iAsIdx = return $ Con $ Coerce ty i
+intToIndexE (RecordTy types) i = do
+  sizes <- traverse indexSetSizeE types
+  (strides, _) <- traverseFunM
+    (\sz prev -> do {v <- imul sz prev; return ((prev, v), v)}) sizes (IntVal 1)
+  offsets <- flip mapM (zip (toList types) (toList strides)) $
+    \(ty, (s1, s2)) -> do
+      x <- irem i s2
+      y <- idiv x s1
+      intToIndexE ty y
+  return $ Record (restructure offsets types)
+intToIndexE rty@(VariantTy types) i = do
+  sizes <- traverse indexSetSizeE types
+  (offsets, _) <- traverseFunM
+    (\sz prev -> do {v <- iadd sz prev; return (prev, v)}) sizes (IntVal 0)
+  let
+    reflect = reflectLabels types
+    -- Find the right index by looping through the possible offsets
+    -- TODO: is there a better way than generating a case statement? Select
+    -- doesn't work because variants aren't primitive types
+    go prev ((label, repeatNum), ty, offset) = do
+      beforeThis <- unsafeCoerce PreludeBoolTy =<< ilt i offset
+      hereCase <- buildNAbs Empty $ \[] -> do
+        shifted <- isub i offset
+        index <- intToIndexE ty shifted
+        return $ Variant types label repeatNum index
+      beforeCase <- buildNAbs Empty $ \[] -> return $ prev
+      emit $ Case beforeThis [hereCase, beforeCase] rty
+    ((l0, 0), ty0, _):zs = zip3 (toList reflect) (toList types) (toList offsets)
+  start <- Variant types l0 0 <$> intToIndexE ty0 i
```

Comment:
Isn't this just `intToIndexE rty 0`? I understand if you want to make it more efficient, but it might deserve a comment

Reply:
Sorry, I don't follow? This is in the definition of `intToIndexE rty i` so I don't think I want to recursively call `intToIndexE rty` here. Also the offset isn't necessarily 0. This line is essentially saying "if the first variant is the correct variant to use, what would the index atom be", and then the `foldM` overrides this if it wasn't actually the first variant.

Maybe I'm not understanding the suggestion?